segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeepertier2/BookkeeperStorage.java

```diff
@@ -0,0 +1,221 @@
+ *
+ * 1. Check the current epoc for the given StorageLedger.
+ * 2. If the current epoc is larger, set it to the new epoc, other wise throw StorageNotPrimary error.
+ * 3. Try and fence all the ledgers. The last ledger may be empty, in this case delete the ledger.
+ * 4. Create a new ledger and add it to the list of ledgers as well as to the ZK.
+ *
+ * Concat: Apache bookkeeper does not have a native concat which appends two ledgers on the server side. To overcome this,
+ * concatenation is implemented as a pure metadata operation. Here is the algorithm used for concat() call:
+ *
+ * 1. Get the target and source ledgers.
+ * 2. Update the offset of the source ledgers to the offsets after the concatenation.
+ * 3. Update all the metadata about ledgers in one transaction using transaction() API for curator.
+ */
+@Slf4j
+public class BookkeeperStorage implements Storage {
+    private static final int NUM_RETRIES = 3;
+
+    //region members
+
+    private final BookkeeperStorageConfig config;
+    private final ExecutorService executor;
+    private final AtomicBoolean closed;
+    private final StorageLedgerManager manager;
+    private final CuratorFramework zkClient;
+
+    //endregion
+
+    //region constructor
+
+    /**
+     * Creates a new instance of the BookkeeperStorage class.
+     *
+     * @param config   The configuration to use.
+     * @param zkClient Curator framework to interact with ZK.
+     * @param executor The executor to use for running async operations.
+     */
+    public BookkeeperStorage(BookkeeperStorageConfig config, CuratorFramework zkClient, ExecutorService executor) {
+        Preconditions.checkNotNull(config, "config");
+        Preconditions.checkNotNull(executor, "executor");
+        this.closed = new AtomicBoolean(false);
+        this.config = config;
+        this.zkClient = zkClient;
+        this.executor = executor;
+        manager = new StorageLedgerManager(config, zkClient, executor);
+    }
+
+    //endregion
+
+    //region Storage implementation
+
+    /**
+     * Initialize is a no op here as we do not need a locking mechanism in case of file system write.
+     *
+     * @param containerEpoch The Container Epoch to initialize with (ignored here).
+     */
+    @Override
+    public void initialize(long containerEpoch) {
+        manager.initialize(containerEpoch);
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openRead(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openRead", streamSegmentName);
+
+        return manager.exists(streamSegmentName, null).thenApply(exist -> {
+            if (exist) {
+                return BookkeeperSegmentHandle.readHandle(streamSegmentName);
+            } else {
+                throw new CompletionException(new StreamSegmentNotExistsException(streamSegmentName));
+            }
+        }).exceptionally((Throwable exception) -> {
```

Comment:
Why are we converting any exception into `StreamSegmentNotExistsException`?

Reply:
Removed the exceptionally clause.